test/common/TestCommon.hs

```diff
@@ -0,0 +1,413 @@
+{-# LANGUAGE DataKinds, RankNTypes, TypeFamilies #-}
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
+module TestCommon ((+\), (*\), (**\),
+                   dReverse0, dReverse1, fX, fX1Y, fXXY, fXYplusZ, fXtoY, freluX,
+                   vec_omit_scalarSum_aux, sumElementsV, altSumElementsV,
+                   sinKonst, sinKonstOut, sinKonstDelay, sinKonstS, sinKonstOutS, sinKonstDelayS,
+                   powKonst, powKonstOut, powKonstDelay,
+                   listsToParameters,
+                   qcPropDom, quickCheckTest0, fquad, quad,
+                   atanReadmeM, atanReadmeDReverse,
+                   vatanReadmeM, vatanReadmeDReverse,
+                  ) where
+
+import Prelude
+
+import qualified Data.Array.DynamicS as OT
+import qualified Data.Array.ShapedS as OS
+import qualified Data.Strict.Vector as Data.Vector
+import qualified Data.Vector.Generic as V
+import           Numeric.LinearAlgebra (Vector)
+import qualified Numeric.LinearAlgebra as HM
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+
+import HordeAd hiding (sumElementsVectorOfDual)
+import HordeAd.Core.DualClass (IsPrimal, dAdd, dScale)
+
+-- Unfortunately, monadic versions of the operations below are not
+-- polymorphic over whether they operate on scalars, vectors or other types,
+-- so we should probably abandon them.
```

Comment:
Hah, this whole comment is, fortunately, no longer true and can be removed (and in another file, unless it's been factored out).